File: pumper/web/src/eqMath.test.ts
import { describe, expect, it } from "vitest";
import { calculateAutoPreamp, compositeGainDb, responseCurve } from "./eqMath";
import { defaultConfig, EqConfig, FilterType, WidthMode } from "./protocol";

function flatConfig(): EqConfig {
  return {
    enabled: true,
    preampDb: 0,
    bands: defaultConfig.bands.map((band) => ({ ...band, enabled: false })),
  };
}

describe("EQ response math", () => {
  it("keeps a flat response at zero decibels", () => {
    const config = flatConfig();
    expect(compositeGainDb(config, 48000, 1000)).toBeCloseTo(0, 8);
    expect(responseCurve(config, 48000, 8).every((point) => Math.abs(point.gainDb) < 1e-8)).toBe(true);
  });

  it("keeps a flat auto preamp at zero decibels", () => {
    expect(calculateAutoPreamp(flatConfig(), 48000)).toEqual({ preampDb: 0, peakDb: 0 });
  });

  it("accounts for overlapping filter gains", () => {
    const config = flatConfig();
    config.bands[0] = {
      enabled: true,
      type: FilterType.Peaking,
      widthMode: WidthMode.Q,
      frequencyHz: 1000,
      gainDb: 6,
      q: 2,
      bandwidthOctaves: 1,
    };
    config.bands[1] = { ...config.bands[0] };
    const result = calculateAutoPreamp(config, 48000);
    expect(result.peakDb).toBeCloseTo(12, 2);
    expect(result.preampDb).toBeCloseTo(-12, 2);
  });

  it("does not add positive gain when every enabled filter cuts", () => {
    const config = flatConfig();
    config.bands[0] = { ...defaultConfig.bands[0], gainDb: -6 };
    expect(calculateAutoPreamp(config, 48000).preampDb).toBeLessThanOrEqual(0);
  });

  it("matches the firmware response for gain-independent filters at every supported rate", () => {
    const rates = [44100, 48000, 88200, 96000, 176400, 192000];
    for (const sampleRate of rates) {
      const config = flatConfig();
      const base = { ...defaultConfig.bands[0], enabled: true, widthMode: WidthMode.Q, frequencyHz: 1000, gainDb: 0, q: 0.707 };

      config.bands[0] = { ...base, type: FilterType.LowPass };
      expect(compositeGainDb(config, sampleRate, 100)).toBeGreaterThan(-0.1);
      expect(compositeGainDb(config, sampleRate, 10000)).toBeLessThan(-30);

      config.bands[0] = { ...base, type: FilterType.HighPass };
      expect(compositeGainDb(config, sampleRate, 100)).toBeLessThan(-30);
      expect(compositeGainDb(config, sampleRate, 10000)).toBeGreaterThan(-0.1);

      config.bands[0] = { ...base, type: FilterType.Notch };
      expect(compositeGainDb(config, sampleRate, 1000)).toBeLessThan(-100);

      config.bands[0] = { ...base, type: FilterType.BandPass };
      expect(compositeGainDb(config, sampleRate, 1000)).toBeCloseTo(0, 6);
      expect(responseCurve(config, sampleRate, 128).every((point) => Number.isFinite(point.gainDb))).toBe(true);
    }
  });

  it("treats filters at or above Nyquist as neutral", () => {
    const config = flatConfig();
    config.bands[0] = { ...defaultConfig.bands[0], enabled: true, type: FilterType.LowPass, widthMode: WidthMode.Q, frequencyHz: 20000, gainDb: 0, q: 0.707 };
    expect(compositeGainDb(config, 32000, 1000)).toBe(0);
  });
});

File: pumper/android/app/src/test/java/dev/wolf/yuan/pumper/protocol/EqMathTest.kt
package dev.wolf.yuan.pumper.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class EqMathTest {
    @Test
    fun `flat response and preamp remain zero`() {
        val curve = EqMath.responseCurve(DefaultEqConfig, 48_000)
        assertTrue(curve.all { kotlin.math.abs(it.gainDb) < 0.000001 })
        assertEquals(0.0, EqMath.calculateAutoPreamp(DefaultEqConfig, 48_000).preampDb, 0.000001)
    }

    @Test
    fun `auto preamp accounts for overlapping gains`() {
        val boosted = DefaultEqConfig.copy(
            bands = DefaultEqConfig.bands.mapIndexed { index, band ->
                if (index < 2) band.copy(type = FilterType.Peaking, widthMode = WidthMode.Q, frequencyHz = 1000.0, gainDb = 6.0, q = 1.0)
                else band
            },
        )
        val result = EqMath.calculateAutoPreamp(boosted, 48_000)
        assertTrue(result.peakDb > 11.9)
        assertTrue(result.preampDb < -11.9)
    }

    @Test
    fun `disabled EQ ignores preamp and filters`() {
        val config = DefaultEqConfig.copy(enabled = false, preampDb = 12.0)
        assertEquals(0.0, EqMath.compositeGainDb(config, 48_000, 1000.0), 0.0)
        assertEquals(0.0, EqMath.responseCurve(config, 48_000).first().gainDb, 0.0)
    }

    @Test
    fun `new RBJ filters have expected response at every supported rate`() {
        val rates = listOf(44_100L, 48_000L, 88_200L, 96_000L, 176_400L, 192_000L)
        rates.forEach { rate ->
            val lowPass = configWith(FilterType.LowPass, 1_000.0)
            val highPass = configWith(FilterType.HighPass, 1_000.0)
            val notch = configWith(FilterType.Notch, 1_000.0)
            val bandPass = configWith(FilterType.BandPass, 1_000.0)

            assertEquals(-3.01, EqMath.compositeGainDb(lowPass, rate, 1_000.0), 0.08)
            assertEquals(-3.01, EqMath.compositeGainDb(highPass, rate, 1_000.0), 0.08)
            assertTrue(EqMath.compositeGainDb(lowPass, rate, 100.0) > -0.1)
            assertTrue(EqMath.compositeGainDb(highPass, rate, 100.0) < -35.0)
            assertTrue(EqMath.compositeGainDb(notch, rate, 1_000.0) < -100.0)
            assertEquals(0.0, EqMath.compositeGainDb(bandPass, rate, 1_000.0), 0.0001)
            FilterType.entries.drop(3).forEach { type ->
                assertTrue(EqMath.responseCurve(configWith(type, 1_137.0), rate).all { it.gainDb.isFinite() })
            }
        }
    }

    @Test
    fun `new filters ignore gain but contribute to auto preamp and bypass at nyquist`() {
        val lowPass = configWith(FilterType.LowPass, 1_000.0, gainDb = 18.0)
        val neutralGain = configWith(FilterType.LowPass, 1_000.0, gainDb = 0.0)
        assertEquals(
            EqMath.compositeGainDb(neutralGain, 48_000, 4_000.0),
            EqMath.compositeGainDb(lowPass, 48_000, 4_000.0),
            0.000001,
        )
        assertTrue(EqMath.calculateAutoPreamp(configWith(FilterType.BandPass, 1_000.0), 48_000).peakDb > -0.01)
        assertEquals(0.0, EqMath.compositeGainDb(configWith(FilterType.LowPass, 24_000.0), 48_000, 8_000.0), 0.0)
    }

    private fun configWith(type: FilterType, frequencyHz: Double, gainDb: Double = 0.0): EqConfig = EqConfig(
        enabled = true,
        preampDb = 0.0,
        bands = listOf(
            EqBand(true, type, WidthMode.Q, frequencyHz, gainDb, 1.0 / kotlin.math.sqrt(2.0), 1.0),
        ),
    )
}
